src/providers/okta.ts

```diff
@@ -0,0 +1,104 @@
+import { OAuth2Client } from "oslo/oauth2";
+import type { OAuth2ProviderWithPKCE } from "../index.js";
+import { TimeSpan, createDate } from "oslo";
+
+export class Okta implements OAuth2ProviderWithPKCE {
+	private client: OAuth2Client;
+	private clientSecret: string;
+
+	constructor(
+		oktaDomain: string,
+		clientId: string,
+		clientSecret: string,
+		redirectURI: string,
+		options: {
+			authServerId?: string;
+		}
+	) {
+		const authorizeEndpoint = `${oktaDomain}/oauth2/${
+			options.authServerId ? `${options.authServerId}/` : ""
+		}v1/authorize`;
+		const tokenEndpoint = `${oktaDomain}/oauth2/${
+			options.authServerId ? `${options.authServerId}/` : ""
+		}v1/token`;
+		this.client = new OAuth2Client(clientId, authorizeEndpoint, tokenEndpoint, {
+			redirectURI
+		});
+		this.clientSecret = clientSecret;
+	}
+
+	public async createAuthorizationURL(
+		state: string,
+		codeVerifier: string,
+		options?: {
+			scopes?: string[];
+		}
+	): Promise<URL> {
+		const url = await this.client.createAuthorizationURL({
+			codeVerifier,
+			scopes: [...(options?.scopes ?? []), "openid"]
+		});
+		url.searchParams.set("state", state);
+
+		return url;
+	}
+
+	public async validateAuthorizationCode(code: string, codeVerifier: string): Promise<OktaTokens> {
+		const result = await this.client.validateAuthorizationCode<TokenResponseBody>(code, {
+			authenticateWith: "request_body",
+			codeVerifier,
+			credentials: this.clientSecret
+		});
+
+		const tokens: OktaTokens = {
+			accessToken: result.access_token,
+			refreshToken: result.refresh_token ?? null,
+			accessTokenExpiresAt: createDate(new TimeSpan(result.expires_in, "s")),
+			idToken: result.id_token,
+			deviceSecret: result.device_secret ?? null
+		};
+
+		return tokens;
+	}
+
+	public async refreshAccessToken(
+		refreshToken: string,
+		options?: {
+			scopes?: string[];
+		}
+	): Promise<OktaTokens> {
+		const result = await this.client.refreshAccessToken<TokenResponseBody>(refreshToken, {
+			authenticateWith: "request_body",
+			credentials: this.clientSecret,
+			scopes: options?.scopes ?? []
```

Comment:
If we always include `openid`, this should be updated

Reply:
If you pass no scopes when refreshing okta will reuse the scopes from the original access_token. If I pass `openid` there it would force people to pass scopes since passing one scope opts out of the "use last" thing.

Reply:
Ah, makes sense. Thought it was a required and that you had to manually define them. Is there a reason to have the option then 🤔

Reply:
Yea, their docs do not make this clear and I had to find this out through testing.

The only reason I would keep the option is because they do support sending the scopes on their endpoint.

So you can essentially choose what you need each time you refresh as long as it's a subset of the scopes used the first time.

I must be honest I can't exactly see a use case for it but they support it and so I wanted to match that, would definitely not be against removing it for the sake of simplicity 😅

Reply:
Do you get the ID token even when you omit the `openid` scope?

> To refresh your access token and an ID token, you send a token request with a `grant_type` of `refresh_token`. Be sure to include the `openid` scope when you want to refresh the ID token.

https://developer.okta.com/docs/guides/refresh-tokens/main/#use-a-refresh-token

Reply:
As far as I remember that was the case. But now that you ask I am slightly unsure, will do another test tonight to make sure. If not we will probably want to include it implicitly of course.

Reply:
Thanks!

Reply:
Have now tested and can confirm that when passing no scopes the ID token is updated with new data on a refresh. Tried changing my users name and the ID token immediately reflected the chaneg after a refresh!